Create prototype for container CVE remediation

querti opened the pull request:
Introduce a new class ContainerVulnerabilities responsible for adding CVE information to container PRs. It's written analogously to the Vulnerabilities class responsible for CVE information in OSV ecosystems. The reason why the new functionality couldn't be combined with the existing class is that it needs to be called after the updates have been resolved (can be seen in extract-update.ts). The CVE evaluation logic is also different from version-based OSV ecosystems. Lastly, creating a new file better separetes the new logic from upstream, which will make rebasing easier.

The evaluation logic is as follows:
For each detected dependency update:
1. If the update is not a Docker type or the image is not specified via digest, skip
2. Gather timestamps of when the current and updated images were created (using Docker API)
3. Gather all vulnerabilities affecting the repo of the updated image
4. Only keep vulnerabilities that are newer than the old image and older than the new image (this implies that old image was affected by a vulnerability and the new image fixes it)
5. Create package rules based on vulnerabilities that remain
6. Package rules will apply the CVE information to the PRs

In its current state, the new module can gather relevant dependency updates, get the 'created' timestamps, filter vulnerabilities based on timestamps, and create packageRules and PR notes. It is NOT able to correctly gather the vulnerabilities affecting a given repository as this functionality is not yet implemented. For this reason, there is a hardcoded vulnerability in the code that was used for manual testing.

Existing methods in the Docker datasource are used to get the image config digest and the config manifest. However, small changes were made: getConfigDigest is no longer private to be usable in a different module, and getImageConfig was reimplemented as getImageConfigFull, since the original method returns only a truncated config.

The full renovate test suite is verified to pass, so the new functionality doesn't cause any regressions in the existing code.

querti committed:
Create prototype for container CVE remediation

Introduce a new class ContainerVulnerabilities responsible for adding
CVE information to container PRs. It's written analogously to the
Vulnerabilities class responsible for CVE information in OSV ecosystems.
The reason why the new functionality couldn't be combined with the
existing class is that it needs to be called after the updates have been
resolved (can be seen in extract-update.ts). The CVE evaluation logic is
also different from version-based OSV ecosystems. Lastly, creating a new
file better separetes the new logic from upstream, which will make
rebasing easier.

The evaluation logic is as follows:
For each detected dependency update:
1. If the update is not a Docker type or the image is not specified via
   digest, skip
2. Gather timestamps of when the current and updated images were created
   (using Docker API)
3. Gather all vulnerabilities affecting the repo of the updated image
4. Only keep vulnerabilities that are newer than the old image and older
   than the new image (this implies that old image was affected by a
   vulnerability and the new image fixes it)
5. Create package rules based on vulnerabilities that remain
6. Package rules will apply the CVE information to the PRs

In its current state, the new module can gather relevant dependency
updates, get the 'created' timestamps, filter vulnerabilities based on
timestamps, and create packageRules and PR notes. It is NOT able to
correctly gather the vulnerabilities affecting a given repository as
this functionality is not yet implemented. For this reason, there is a
hardcoded vulnerability in the code that was used for manual testing.

Existing methods in the Docker datasource are used to get the image
config digest and the config manifest. However, small changes were made:
getConfigDigest is no longer private to be usable in a different module,
and getImageConfig was reimplemented as getImageConfigFull, since the
original method returns only a truncated config.

The full renovate test suite is verified to pass, so the new
functionality doesn't cause any regressions in the existing code.

qixiang commented on lib/workers/repository/process/container-vulnerabilities.ts:
@querti do you know if security team will provide such data (in CSAF or OSV format), or we need to convert it from regular CSAF files for container security advisories?

querti replied:
I actually don't know what format the provided data will be in, but my plan is to convert it to OSV (or very similar) format in MintMaker  - same as RPMs. If necessary, we may also do some data transformation, for example getting a list of affected repos and adding a new field with this value into the OSV format. It may help us filter relevant advisories that affect a given repo. I'd like to extend our fork of osv-offline to be able to handle this functionality.

qixiang replied:
Could you use some fake data to verify whether the osv-offline queries can work with data that doesn't have
`ranges` at all? You don't have to test that with this container logic, just verify if it can support such queries for existing supported packages.

This is how the regular osv data looks like:
```
"affected": [
  {
    "package": {
      "name": "github.com/strukturag/libheif",
      "ecosystem": "Go",
      "purl": "pkg:golang/github.com/strukturag/libheif"
    },
    "ranges": [
      {
        "type": "SEMVER",
        "events": [
          {
            "introduced": "0"
          },
          {
            "fixed": "1.15.2"
          }
        ]
      }
    ]
  }
]
```
The `fixed` data is provided to perform the queries, I'm not sure if we can do queries that just check if a package (a repository in this case) is affected by any cve advisory without providing the package version.

querti replied:
Would you mind if I did it as a part of a follow-up Jira where I implement container advisory support in the osv-offline database? Regardless of what the result might be, I don't think it will have a large impact on what I have implemented so far. And this PR is large enough already so I would prefer not to increase its scope further.

qixiang replied:
yeah, no problem, since you're merging this into a dedicate branch, personally I think it's fine to just merge this initial version, once we have a ready-to-work version, we can squash the changes and prepare a better version. Please also create jira issues to track remaining work.

staticf0x commented on lib/workers/repository/process/container-vulnerabilities.ts:
```diff
@@ -0,0 +1,478 @@
+  ): Promise<ContainerVulnerability[]> {
+    const managers = Object.keys(packageFiles);
+
+    // TODO: should we also include other docker managers: devcontainer, docker-compose ?
```
We don't yet support those in MintMaker but we support others like tekton and even a custom regex manager can work with docker datasource/versioning.

querti replied:
Good point, I changed the logic to process dependencies with docker datasource instead of docker manager. This way all managers with container updates can receive CVE alerts.

staticf0x commented:
Very nice PR, thank you. Except for that comment above I don't have anything.

querti committed:
Use Container CVE scanning on all dependencies with Docker datasource

Use Docker datasource instead of only allowing Docker manager
dependencies. THere are container image updates in other managers that
would otherwise be missed.

querti commented:
I'll merge this PR since I haven't received any more comments.
